### ISSUE TITLE
Unable to get Hand data

### ISSUE THREAD

**Issue by ValemVR** [handzlikchris/Unity.QuestRemoteHandTracking#7]
Hi Chris,

This project seems amazing however I can't make it work at the moment. I made sure to follow all the steps but I get a "Unable to get Hand data" once I launch the app. 

However, I think that I might not understand the "Running" part. Should I launch the app with Oculus Link on ? Should I build the app on my Quest first and play it in parallel in Unity aftewards? 

Thank you for your time. :)

**Comment by handzlikchris** [comment]
Hi ValemVR,

Thanks, you have to have both running
- build app on Quest with HandTracking and run it there
- then also hit play in Unity Editor

Quest will then send data in real time over TCP and UnityEditor will pick it up in play mode and render hands in similar way as quest would.

It's easy to think of it as Quest being just data feeder. Don't turn on link as that'll not allow you to use hand tracking at this stage.

**Comment by ValemVR** [comment]
Thank you Chris for your help. :)

I will have a go at this.

Have a nice day,

Quentin

**Comment by handzlikchris** [comment]
HTH, let me know how you got on

**Comment by upstanding** [comment]
Hi, I have setup a test app using the train scene demo and I am seeing this same error message in the console as well.  At first I was getting some other exceptions, so I added a few error checks (which I'd be glad to share).  Now my test app on Quest is talking to the Unity editor and I see some initialization packets come through HandsDataReceiver.HandleIncomingData().  When the Quest first sees my hands, the editor receives 2 Skeleton packets and 2 Mesh packets and I see the 4 "initialized" log messages, but after that it only receives periodic empty (null) packets, presumably from the keepalives.  I never see the hands appear in the editor Scene view.  

Additional information which may be useful:
- The "Unable to get hand data" error appears once immediately on startup.
- The Quest USB cable is disconnected from the PC while running the test app.
- The Quest app freezes on a gray screen on startup if it cannot connect to the Unity editor, but it usually resumes a few seconds after I press play in the editor.
- The Quest app also freezes while the Unity editor is paused at a breakpoint.
- The hands are working properly in the train scene test app in the headset.

Any ideas on what the issue could be?  Is something causing the Quest app to not send update packets to the editor?  Thanks for your help.

**Comment by handzlikchris** [comment]
Hi Upstanding,

Thanks for info. If you can share those additional error checks I'll add them in.

It's interesting that it freezes in Quest, I've not been able to recreate that even when network was off.

If it managed to send that initialisation data then it's nearly there. Hand data is actually sent via UDP rather than TCP (which is used for skeleton/mesh initialisation).

- is your firewall not blocking UDP? 
- can you share logcat logs from quest, please
it should write some information if hand data is passed through, like
```
01-15 09:59:38.406: I/Unity(5191): UDP Connected hands data endpoint
01-15 09:59:44.151: I/Unity(5191): Message sent (size: 2948)
```
- if the messages are sent from device it could be something network related, UDP packages are not guaranteed to be delivered (although they're fast and Quest sends 100s per second so if few are lost it should not cause issues)

**Comment by upstanding** [comment]
After trying several things, I switched to the setup without the dependency on SimpleFramework and I was able to get it working correctly.  Thanks again.

I have attached a zip that includes three updated files with some checks to prevent certain exceptions when I was using the OculusSimpleFramework version.  The changes to HandsDataReceiver and XMLSerialize were to deal with the empty packets I was receiving, though it could be a bit more optimized.  The change to TcpReciever prevents an exception from being thrown on close.

[Edits.zip](https://github.com/handzlikchris/Unity.QuestRemoteHandTracking/files/4120733/Edits.zip)
